source/common/network/connection_impl.h

```diff
@@ -72,11 +72,16 @@ class ConnectionImpl : public virtual Connection,
   void write(Buffer::Instance& data) override;
   void setReadBufferLimit(uint32_t limit) override { read_buffer_limit_ = limit; }
   uint32_t readBufferLimit() const override { return read_buffer_limit_; }
+  void setWriteBufferWatermarks(size_t low_watermark, size_t high_watermark) override;
 
   // Network::BufferSource
   Buffer::Instance& getReadBuffer() override { return read_buffer_; }
   Buffer::Instance& getWriteBuffer() override { return *current_write_buffer_; }
 
+  void replaceWriteBufferForTest(std::unique_ptr<Buffer::OwnedImpl> new_buffer) {
+    write_buffer_ = std::move(new_buffer);
```

Comment:
Agreed. The general pattern we have been doing is to make something like `BufferFactory` which returns a buffer. Then in a test you can hold on to the pointer to do things. Search for "Factory" and you will see many examples.